### PR TITLE
driver: Improvements around binary locating

### PR DESCRIPTION
**Pull request by nathanchance**
We've tailored driver.sh around our Docker image with regards to the
binary versions that we check for. A user might not have the latest
versions of Clang or llvm-ar installed. Check for all three versions
offered by apt.llvm.org, falling back to the generic binary name if
those are not found.

Additionally, for llvm-ar, do a sanity check that the user has a version
of llvm-ar that can actually build the kernel (has support for the P
modifier). If they don't fall back to regular GNU ar and let them know.

Presubmit: https://travis-ci.com/nathanchance/continuous-integration/builds/103470578

### CODE REVIEW

File: driver.sh [hunk 1]
```diff
@@ -155,11 +193,12 @@ mako_reactor() {
...
   KBUILD_BUILD_USER=driver \
   KBUILD_BUILD_HOST=clangbuiltlinux \
-  make -j"${jobs:-$(nproc)}" CC="${CC}" HOSTCC="${CC}" LD="${LD}" HOSTLD="${HOSTLD:-ld}" AR="llvm-ar-9" "${@}"
+  make -j"${jobs:-$(nproc)}" CC="${CC}" HOSTCC="${CC}" LD="${LD}" HOSTLD="${HOSTLD:-ld}" AR="${AR}" "${@}"
 }
 
 build_linux() {
-  CC="$(command -v ccache) $(command -v clang-9)"
+  # Wrap CC in ccache
+  CC="$(command -v ccache) ${CC}"
```
Comment (nickdesaulniers) [inline driver.sh]: can we move this ccache check above?

File: driver.sh [hunk 2]
```diff
@@ -139,14 +140,51 @@ setup_variables() {
...
+  if [[ -z "${AR:-}" ]]; then
+    for AR in llvm-ar-9 llvm-ar-8 llvm-ar-7 llvm-ar "${CROSS_COMPILE:-}"ar; do
+      ${AR} --version 2>/dev/null && break
+    done
+    # Optimistically check to see that the user has a llvm-ar
+    # with https://reviews.llvm.org/rL354044. If they don't,
+    # fall back to GNU ar and let them know.
+    if ${AR} --version | grep -q "LLVM" && \
+       [[ $(${AR} --version | grep version | sed -e 's/.*LLVM version //g' -e 's/[[:blank:]]*$//' -e 's/\.//g') -lt 900 ]]; then
+      set +x
```
Comment (nickdesaulniers) [inline driver.sh]: Can you put this check in another function with a descriptive name like `check_ar_version` or w/e?

File: driver.sh [hunk 3]
```diff
@@ -139,14 +140,51 @@ setup_variables() {
...
+  # don't exist, try to find them.
+  # lld isn't ready for all architectures so it's just
+  # simpler to fall back to GNU ld when LD isn't specified
+  # to avoid architecture specific selection logic.
+
+  "${LD:="${CROSS_COMPILE:-}"ld}" --version
+
+  if [[ -z "${CC:-}" ]]; then
+    for CC in clang-9 clang-8 clang-7 clang; do
+      ${CC} --version 2>/dev/null && break
```
Comment (nickdesaulniers) [inline driver.sh]: what happens when no version is found?

File: driver.sh (hunk 1 above)
Comment (nathanchance) [reply driver.sh]: `ccache` isn't a required binary, that's why it is here instead of up there. I'll make that a comment.

File: driver.sh (hunk 2 above)
Comment (nathanchance) [reply driver.sh]: Yes.

File: driver.sh (hunk 3 above)
Comment (nathanchance) [reply driver.sh]: Hmmm fair point, I'll need to add another check to make sure we fail here instead of during make.

**Comment by nathanchance** [comment]
All comments should be addressed. If no `${CC}` or `${AR}` is found, the version command will now properly fail because we `set -e` at the top.

**Comment by nathanchance** [comment]
I went ahead and added a more descriptive message if `clang` is not found (or if the user specified a non-existent binary in PATH).

**Pull request merged**